### ISSUE TITLE
S3 as destination

### ISSUE THREAD

**Issue by deepakksahu** [airbytehq/airbyte#573]
## Tell us about the new integration you’d like to have
AWS s3 is a heavily used destination for the data lake. If we can sink to that that would be great.

## Describe the context around this new integration
Data Engineering team does the daily ingestion and all the time the sync is S3 for me.

## Describe the alternative you are considering or using
Custom spark program.

┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1200367912513076/1200368234046956) by [Unito](https://www.unito.io)

**Comment by sherifnada** [comment]
@deepakksahu thanks for making this request! Is there a specific data format you want to send to S3 (e.g: CSV), or are there different formats you would be interested in sending?

**Comment by deepakksahu** [comment]
One of the most used file format for the OLAP will be parquet,orc format. In some cases it can be AVRO or CSV(rare) as well.

According to me if we can support all 4 that will be great. Otherwise can start with first two.

**Comment by sherifnada** [comment]
We'll prioritize S3 as a destination in our roadmap shortly. Thanks for the clarification!

**Comment by rubenssoto** [comment]
Hello guys,

Parquet is a good one. :)

**Comment by dnskr** [comment]
Hi!
Delta Lake format is very good and popular candidate: [Home](https://delta.io) and [Docs](https://docs.delta.io/latest/index.html)

**Comment by brucemen711** [comment]
Anyone working on Delta lake destination ? We are very excited about this.

**Comment by sherifnada** [comment]
hi @kyprifog! This is not on Airbyte's roadmap for April, but this is definitely top of mind for us and will update you as we form our May roadmap. We'll also create some documentation for how to create a destination yourself in case you need to unblock yourself for the time being.

**Comment by kyprifog** [comment]
@sherifnada any update on that documentation?

**Comment by sherifnada** [comment]
@kyprifog we'll have it this week!

**Comment by eduardgruy** [comment]
+1. Would be great to have S3 destination that supports multiple formats. Ideally Deltalake/CSV/Parquet. 
I suppose S3 would be used as a kind of landing zone so appending changes to a delta lake table would be perfect.

**Comment by sherifnada** [comment]
Hi everyone! We're going to start work on the S3 destination very soon and would love to hear your feedback on features you need. Could you comment on this thread with: 

* What volume of data do you anticipate replicating? bytes or number of records estimate is helpful to know. 
* Data format you'd like to write in S3
* Does the data you write need to be partitioned? What partitioning scheme do you need?
* How are you using the data in S3? Are you importing it into Redshift, athena, dashboards, etc..

**Comment by sherifnada** [comment]
Hi everyone! We're going to start work on the S3 destination very soon and would love to hear your feedback on features you need. Could you comment on this thread about any of the following:
* What volume of data do you anticipate replicating? bytes or number of records estimate is helpful to know.
* Data format you'd like the connector to support
* Does the data you write need to be partitioned? What partitioning scheme do you need?
* How are you using the data in S3? Are you importing it into Redshift, athena, dashboards, etc.. 
* What is the UX you want for interacting with the S3 connector? What inputs/configuration do you want to give it, and how do you want the output to look like?

**Comment by mmolimar** [comment]
Hi @sherifnada 
This is great news! My thoughts below :-)  
> Hi everyone! We're going to start work on the S3 destination very soon and would love to hear your feedback on features you need. Could you comment on this thread about any of the following:
> 
> * What volume of data do you anticipate replicating? bytes or number of records estimate is helpful to know.

We've estimated about 5-10 million records on each connection, having dozens connections (based on sync settings).

> * Data format you'd like the connector to support

JSONL, Avro. If possible having something like DeltaLake and "upserts" it'd be great!

> * Does the data you write need to be partitioned? What partitioning scheme do you need?

Yes. Partitioned by stream name, load timestamp plus some custom partitions to be filled by the user.

> * How are you using the data in S3? Are you importing it into Redshift, athena, dashboards, etc..

Yes, using Redshift, Athena, and other downstream apps to process this data.

> * What is the UX you want for interacting with the S3 connector? What inputs/configuration do you want to give it, and how do you want the output to look like?

A form to configure the credentials and the S3 path in which we could include dynamic values in it (something like parsing ``DateTimeFormatter`` patterns to generate dynamic URIs).

**Comment by kyprifog** [comment]
I had very simple use case, I wanted to create a quick and easy sync between google sheets and S3.   There are other ways to do this, but wanted to see if I could knock it out with airbyte since i was using for other things.

**Comment by sherifnada** [comment]
@mmolimar for the initial release we are going to support CSV as the data format, but will quickly start adding other formats like JSONL, Parquet, and Avro. We will also separate release a delta lake connector (#2075 ) in the future to make importing into delta lake a seamless process.

**Comment by cbows** [comment]
Hi @sherifnada

> Data format you'd like the connector to support?

CSV (maybe gzipped). Later support for Parquet, Avro, Delta Lake would be nice.

> Does the data you write need to be partitioned? What partitioning scheme do you need?

Yes. The Hive data model (see [wiki](https://cwiki.apache.org/confluence/display/Hive/Design)) is probably very popular and also the one I use. So it would be nice to have the possibility to specify one or multiple columns in the source that are used to partition the data on S3 (e.g. `s3a://mybucket/myschema/mytable/country=ABC/region=DEF/data.part1.csv`). The `country` and `region` columns don't have to be in the `data.part1.csv` if partitioning is used like that. Additional data could be added in the future using the correct object/partition key (ex. `/country=ABC/region=DEF/data.part2.csv`)

> How are you using the data in S3? Are you importing it into Redshift, Athena, dashboards, etc.?

Data is further transformed and queried with [Trino](https://github.com/trinodb/trino) (f.k.a PrestoSQL) using the Hive connector. I'd like to add the CSV data written by Airbyte as an external table in Hive (see [wiki](https://cwiki.apache.org/confluence/display/Hive/Managed+vs.+External+Tables)). From there on, the data can be queried and transformed, e.g converted to Parquet, directly with Trino (or something like DBT).

> What is the UX you want for interacting with the S3 connector? What inputs/configuration do you want to give it, and how do you want the output to look like?

* Specify S3 credentials
* Specify additional S3 settings like host, region and URL scheme, to be able to use Airbyte with alternative S3-compliant object storage implementations like [MinIO](https://github.com/minio/minio).
* Define partition keys to be used from the source data
* Set compression on/off

**Comment by janhavi-debug** [comment]
Hi @sherifnada, I was planning to build an S3 destination connector myself as well and came across airbyte's Redshift connector that has a logic where it puts data into S3 and then copies it to Redshift from there, so what I can do is reverse engineer the Redshift connector till the S3 task and build it in my DEV version. However, it would be great if you guide me as well. 

Thanks and regards.

**Comment by sherifnada** [comment]
@janhavi-debug no need! we should have a PR for S3 destination by EOW :) Though we'd love it if you help us understand your usecase by answering some of the questions above!

**Comment by olivermeyer** [comment]
Hi @sherifnada, just adding my two cents to some of your questions as we're waiting for this S3 destination to start using Airbyte.

> Data format you'd like the connector to support

Definitely Parquet!

> Does the data you write need to be partitioned? What partitioning scheme do you need?

Yes, and I agree with Hive partitioning. One key feature for us is the ability to use partition values which are not in the data (e.g. we partition by processing time, not event time).

> How are you using the data in S3? Are you importing it into Redshift, athena, dashboards, etc..

Data will be read by Redshift Spectrum.

Thanks and looking forward to the release!

**Comment by tuliren** [comment]
Update on this topic:

1. The S3 destination connector has been merged to master. [Here](https://github.com/airbytehq/airbyte/blob/master/docs/integrations/destinations/s3.md) is the documentation.
2. Currently it only supports the CSV format.
3. We are working on supporting the Parquet format right now (#3642).
4. More formats and configurations will be supported in the future. Feel free to create new tickets for feature and configuration requests.
